containers/src/Data/IntSet/Internal.hs

```diff
@@ -913,22 +916,33 @@ split x t =
 splitMember :: Key -> IntSet -> (IntSet,Bool,IntSet)
 splitMember x t =
   case t of
-      Bin _ m l r | m < 0 -> if x >= 0
-                             then case go x l of
-                                 (lt, fnd, gt) -> let !lt' = union lt r
-                                                  in (lt', fnd, gt)
-                             else case go x r of
-                                 (lt, fnd, gt) -> let !gt' = union gt l
-                                                  in (lt, fnd, gt')
-      _ -> go x t
+    Bin p m l r
+      | m < 0 ->
+        if x >= 0 -- handle negative numbers.
+        then
+          case go x l of
+            (lt, fnd, gt) ->
+              let !lt' = bin p m lt r
+              in (lt', fnd, gt)
+        else
+          case go x r of
+            (lt, fnd, gt) ->
+              let !gt' = bin p m l gt
+              in (lt, fnd, gt')
+    _ -> go x t
   where
     go x' t'@(Bin p m l r)
-        | match x' p m = if zero x' m
-                         then case go x' l of
-                             (lt, fnd, gt) -> (lt, fnd, union gt r)
-                         else case go x' r of
-                             (lt, fnd, gt) -> (union lt l, fnd, gt)
-        | otherwise   = if x' < p then (Nil, False, t') else (t', False, Nil)
+        | nomatch x' p m = if x' < p then (Nil, False, t') else (t', False, Nil)
+        | zero x' m =
+          case go x' l of
+            (lt, fnd, gt) ->
+              let !gt' = bin p m gt r
+              in (lt, fnd, gt')
+        | otherwise =
+          case go x' r of
+            (lt, fnd, gt) ->
+              let !lt' = bin p m l lt
```

Comment:
> Where did you add strictness in splitMember?

`gt'` and `lt'` here.

Reply:
Thanks. Now I see it. Should we add a strict triple type alongside our strict pair type? It wouldn't have to be in this PR.

Reply:
Apparently `Data.Map` has one, and uses it for this very purpose: https://github.com/haskell/containers/blob/aebfe5ed88e37b8fc5373ec033c23d6b40faad61/containers/src/Data/Map/Internal.hs#L3856

However `Data.Set` uses bangs in `splitMember` (like this PR), and `Data.IntMap` defines a different strict triple for `splitLookup`. Might be a good idea to move the strict triple to utils and use it in all these places. Probably better in a separate PR though, yes.